In [1]:
import pandas as pd
import numpy as np

# E-Redes datasets inspection

This notebook provides a sequential inspection of datasets selected from E-REDES, focusing on their structure, data quality, temporal and geographic granularity, value ranges, units, and consistency across datasets.

## **8-total-upac-mensal**: Potencia instalada (no obligatoriamente activa) de unidades de autoconsumo

In [2]:
total_upac_df = pd.read_csv(r"../data_raw/8-total-upac-mensal.csv", sep=';', encoding='utf-8')
total_upac_df.head(5)

,data,ano,mes,distrito,concelho,freguesia,codigo_postal,tipo_de_tecnologia,nivel_de_tensao,escalao_de_potencia_instalada,numero_de_instalacoes,potencia_instalada_upac_kw,coddistrito,codconcelho,codfreguesia,cpes,relacao_instalacoes_por_cpe,relacao_potencia_por_cpe
0,2023-10,2023,10,Viseu,Castro daire,Mões,3600,Solar,BTN,"]0, 4]",26.0,42.55,18,1803,180312,12834.0,0.002026,0.003315
1,2023-10,2023,10,Viseu,Castro daire,Monteiras,3600,Solar,BTN,"]0, 4]",5.0,5.75,18,1803,180314,12834.0,0.000390,0.000448
2,2023-10,2023,10,Viseu,Castro daire,"Mamouros, Alva e Ribolhos",3600,Solar,BTN,"]0, 4]",40.0,82.87,18,1803,180323,12834.0,0.003117,0.006457
3,2023-10,2023,10,Viseu,Castro daire,Picão e Ermida,3600,Solar,BTN,"]0, 4]",14.0,21.11,18,1803,180326,12834.0,0.001091,0.001645
4,2023-10,2023,10,Viseu,Castro daire,Reriz e Gafanhão,3600,Solar,BTN,"]0, 4]",10.0,21.11,18,1803,180327,12834.0,0.000779,0.001645


In [3]:
total_upac_df.shape

(490759, 18)

In [4]:
total_upac_df.dtypes

data                              object
ano                                int64
mes                                int64
distrito                          object
concelho                          object
freguesia                         object
codigo_postal                      int64
tipo_de_tecnologia                object
nivel_de_tensao                   object
escalao_de_potencia_instalada     object
numero_de_instalacoes            float64
potencia_instalada_upac_kw       float64
coddistrito                       object
codconcelho                       object
codfreguesia                      object
cpes                             float64
relacao_instalacoes_por_cpe      float64
relacao_potencia_por_cpe         float64
dtype: object

In [5]:
total_upac_df.describe()

,ano,mes,codigo_postal,numero_de_instalacoes,potencia_instalada_upac_kw,cpes,relacao_instalacoes_por_cpe,relacao_potencia_por_cpe
count,490759.000000,490759.000000,490759.000000,490759.000000,490759.000000,489248.000000,489248.000000,489248.000000
mean,2024.350769,6.186688,4514.633735,18.974678,143.176291,45763.512722,0.000902,0.006841
std,1.040159,3.428097,1672.658093,63.685325,433.313486,61918.446305,0.002632,0.029243
min,2023.000000,1.000000,1000.000000,1.000000,0.000000,1264.000000,0.000002,0.000000
25%,2023.000000,3.000000,3105.000000,1.000000,16.380000,10962.000000,0.000036,0.000542
50%,2024.000000,6.000000,4600.000000,2.000000,35.000000,27057.000000,0.000122,0.001670
75%,2025.000000,9.000000,5180.000000,9.000000,98.120000,58377.000000,0.000523,0.004750
max,2026.000000,12.000000,8970.000000,2389.000000,19600.000000,402927.000000,0.069918,3.176146


In [6]:
total_upac_df.isna().sum()

data                                0
ano                                 0
mes                                 0
distrito                         1511
concelho                         1511
freguesia                        1758
codigo_postal                       0
tipo_de_tecnologia                 21
nivel_de_tensao                    15
escalao_de_potencia_instalada       0
numero_de_instalacoes               0
potencia_instalada_upac_kw          0
coddistrito                         0
codconcelho                         0
codfreguesia                        0
cpes                             1511
relacao_instalacoes_por_cpe      1511
relacao_potencia_por_cpe         1511
dtype: int64

In [7]:
total_upac_df.duplicated().sum()

np.int64(0)

In [8]:
total_upac_df[['ano','mes','codfreguesia', 'codigo_postal', 'tipo_de_tecnologia','nivel_de_tensao','escalao_de_potencia_instalada']].duplicated().sum()

np.int64(0)

In [ ]:
total_upac_df['data'] = pd.to_datetime(total_upac_df['data'], format='%Y')

total_upac_df['data'].dtype, total_upac_df['data'].min(), total_upac_df['data'].max()

(dtype('<M8[ns]'),
 Timestamp('2023-01-01 00:00:00'),
 Timestamp('2026-06-01 00:00:00'))

## Observaciones sobre el dataset '8-total-upac-mensal':


Se comprueba que el dataset presenta 18 columnas, todas ellas con las unidades correctas según lo establecido en el catálogo.
La columna cpes hace referencia al Código do Ponto de Entrega, un identificador único, digital e invariable de 20 dígitos que la distribuidora asigna a cada punto de suministro de energía. Si bien es poco probable que se utilice directamente en este proyecto, era la única variable cuya interpretación faltaba por determinar.

En cuanto al tipo de dato de cada columna, se ha adelantado la transformación de data de texto a datetime para poder determinar correctamente el rango temporal del dataset. Respecto al resto de columnas, los códigos se mantienen como variables identificadoras no numéricas, puesto que no tiene sentido realizar operaciones matemáticas o de agregación sobre ellos. Por otro lado, se podrían transformar a categóricas las columnas tipo_de_tecnologia, nivel_de_tensao y escalao_de_potencia_instalada. En el caso de nivel_de_tensao y escalao_de_potencia_instalada, además, podría establecerse un orden interno si resulta útil para el análisis.

En cuanto a las observaciones que se pueden extraer de la función describe(), se aprecia una dispersión muy elevada en potencia_instalada_upac_kw: mientras que la mediana es de 35 kW y el percentil 75 se sitúa en 98,12 kW, el máximo alcanza los 19.600 kW, lo que indica una distribución fuertemente sesgada hacia valores altos. Antes de considerar estos valores como errores, habría que estudiar si pueden deberse a grandes instalaciones energéticas o a agregaciones de múltiples instalaciones dentro de una misma observación.

Para determinar si estos valores extremos son reales, conviene revisar las filas concretas y comprobar su coherencia con el número de instalaciones, la potencia por instalación, el resto de variables del registro y la evolución temporal de esa misma zona. Un valor extremo puede ser perfectamente válido si responde a una instalación o conjunto de instalaciones de gran tamaño; en cambio, podría considerarse sospechoso si resulta incoherente con el resto de información disponible.

Algo similar sucede con numero_de_instalacoes, que presenta una mediana de 2 instalaciones, un percentil 75 de 9 y un máximo de 2.389 instalaciones. Este valor máximo puede ser perfectamente real si determinadas zonas concentran una gran cantidad de instalaciones UPAC. La elevada dispersión de la potencia instalada también se refleja en relacao_potencia_por_cpe, que presenta igualmente una distribución muy sesgada.

También habría que investigar el hecho de que el mínimo de potencia_instalada_upac_kw sea 0 mientras que el mínimo de numero_de_instalacoes sea 1, ya que esto implicaría la existencia de registros con al menos una instalación pero una potencia total registrada de 0 kW.

En cuanto a los 1.511 valores nulos de distrito, concelho, cpes y las relaciones calculadas a partir de CPE, es muy probable que estén relacionados. Conviene comprobar si los valores faltantes corresponden exactamente a las mismas filas. Dado que los códigos territoriales no presentan valores nulos, en primer lugar se podría intentar recuperar distrito y concelho a partir de dichos códigos. En el caso de cpes, se podría valorar su recuperación mediante registros equivalentes o una imputación por grupo si existe suficiente evidencia. Si no existe una forma fiable de recuperar estos valores, podrían mantenerse como NaN o eliminarse únicamente si interfieren con el análisis.

Los valores nulos de freguesia no parecen especialmente relevantes, ya que el proyecto se enfoca principalmente a nivel de concelho. Por tanto, no sería necesario imputarlos salvo que posteriormente se quiera realizar un análisis específico a nivel de freguesia.

Para los valores faltantes de tipo_de_tecnologia y nivel_de_tensao, aunque son muy pocos en relación con el tamaño total del dataset (21 y 15 respectivamente), se podría realizar una imputación por grupo únicamente si la categoría correspondiente es inequívoca. En caso contrario, sería preferible mantenerlos como NaN o eliminar esas filas si dichas variables son imprescindibles para el análisis.

Respecto a los valores extremos, si se comprueba que son reales se mantendrán. Si se determina que son errores, se intentará recuperar el valor correcto a partir de registros equivalentes o de la fuente original. Si esto no es posible, se valorará una imputación por grupo adecuada o, como última opción, la eliminación del registro.

En cuanto a la granularidad, se comprobó que una observación no se define únicamente por año, mes y freguesia. La combinación también depende de codigo_postal, tipo_de_tecnologia, nivel_de_tensao y escalao_de_potencia_instalada. Además, en algunos registros freguesia aparece como NaN mientras codfreguesia sí está informado, por lo que resulta más fiable utilizar el código territorial para identificar la unidad geográfica.

Por tanto, la granularidad original parece venir determinada por año + mes + codfreguesia + codigo_postal + tipo_de_tecnologia + nivel_de_tensao + escalao_de_potencia_instalada. Posteriormente será necesario agregar estos datos al nivel mes + concelho para combinarlos con el resto de datasets.

## **energia_injectada_upac**: Energía inyectada a la red por instalaciones de autoconsumo (activas)

In [10]:
energia_injectada_df = pd.read_csv(r"../data_raw/energia_injectada_upac.csv", sep=';', encoding='utf-8')
energia_injectada_df.head(5)

,data,ano,mes,ano_mes,distrito,con_name,nivel_tensao,potencia_instalada,numero_de_instalacoes,energia,codigo_distrito,codigo_concelho,codigo_freguesia
0,2025-10,2025,10,202510,Viana do Castelo,Vila Nova de Cerveira,BTN/BTE,980.48,362,34379.180,16,1610,161000
1,2025-06,2025,6,202506,Viseu,Vouzela,BTN/BTE,2183.14,607,134390.782,18,1824,182400
2,2025-12,2025,12,202512,Portalegre,Elvas,MT/AT,4215.74,40,71816.250,12,1207,120700
3,2025-05,2025,5,202505,Portalegre,Elvas,MT/AT,3503.74,36,208800.000,12,1207,120700
4,2026-01,2026,1,202601,Setúbal,Sines,MT/AT,2188.33,12,10461.250,15,1513,151300


In [11]:
energia_injectada_df.shape

(20406, 13)

In [12]:
energia_injectada_df.dtypes

data                      object
ano                        int64
mes                        int64
ano_mes                    int64
distrito                  object
con_name                  object
nivel_tensao              object
potencia_instalada       float64
numero_de_instalacoes      int64
energia                  float64
codigo_distrito           object
codigo_concelho           object
codigo_freguesia          object
dtype: object

In [13]:
energia_injectada_df.describe()

,ano,mes,ano_mes,potencia_instalada,numero_de_instalacoes,energia
count,20406.000000,20406.000000,20406.000000,20406.000000,20406.000000,2.040600e+04
mean,2024.177105,6.180682,202423.891159,3091.770394,414.761051,8.114665e+04
std,0.947170,3.536492,94.066569,4460.974371,782.857078,1.287258e+05
min,2023.000000,1.000000,202301.000000,21.860000,3.000000,0.000000e+00
25%,2023.000000,3.000000,202311.000000,630.092500,13.000000,1.404271e+04
50%,2024.000000,6.000000,202408.000000,1451.465000,98.000000,3.728429e+04
75%,2025.000000,9.000000,202506.000000,3639.875000,428.750000,9.114092e+04
max,2026.000000,12.000000,202603.000000,47345.990000,6173.000000,1.963895e+06


In [14]:
energia_injectada_df.isna().sum()

data                       0
ano                        0
mes                        0
ano_mes                    0
distrito                 368
con_name                 368
nivel_tensao               0
potencia_instalada         0
numero_de_instalacoes      0
energia                    0
codigo_distrito            0
codigo_concelho           39
codigo_freguesia           0
dtype: int64

In [15]:
energia_injectada_df.duplicated().sum()

np.int64(0)

In [16]:
energia_injectada_df[['ano','mes','codigo_concelho','nivel_tensao']].duplicated().sum()

np.int64(0)

In [17]:
energia_injectada_df['data'] = pd.to_datetime(energia_injectada_df['data'])

energia_injectada_df['data'].min(), energia_injectada_df['data'].max()

(Timestamp('2023-01-01 00:00:00'), Timestamp('2026-03-01 00:00:00'))

## Observaciones sobre el dataset 'energia_injectada_upac'


Se confirma que las unidades de cada variables coinciden con las indicadas en el catálogo.

Se comprueba que el dataset tiene una extensión de 20406 filas y 13 columnas, hay que tenerlo en cuenta a la hora de combinarlo con otros datasets con estructura distinta.

La granularidad se determina por año, mes, concelho y nivel de tensión, puesto que una misma combinación de año, mes y concelho se subdivide según 'nivel_tensao'.

Cabe destacar que si bien la columna 'energia' es la que muestra la energía inyectada a la red por instalaciones de autoconsumo, también se presentan columnas que presentan datos compartidos con el anterior dataset tal como las de potencia instalada o número de instalaciones.

Se observa de nuevo que la variable 'data' tiene un tipo erróneo que hay que pasar a datetime para poder calcular correctamente el rango de fechas que abarca el dataset.
La columna 'nivel_tensao' debería ser categórica, el resto de columnas presentan tipos de datos correctos.

De forma similar al dataset anterior, se observa cierta desviación en cuanto a la distribución del número de instalaciones, la potencia instalada y la energía inyectada, se mantiene la hipótesis a comprobar sobre regiones con mayor concentración de instalaciones o potencias más elevadas que otras.

En cuanto a valores nulos, se observa la misma cantidad de nulos para distrito que para nombre de concelho, hay posibilidades reales de sacarlos a partir de 'codigo_concelho' que también presenta nulos pero muchos menos. Habría que investigar si hay formas de recuperar los 39 nulos relativos a dichos códigos dada su potencial utilidad para hacer "merge" con este dataset.

Para identificar territorialmente los registros se priorizará codigo_concelho frente al nombre del concelho, ya que es una clave más estable y además será más útil para realizar futuros merge entre datasets.

Parece que tampoco se encuentras duplicados en este dataset y su rango de fechas es igual al anterior, abarcando desde enero de 2023 a marzo de 2026. Abarca 3 meses menos que el dataset anterior, a tener en cuenta.

## **26-centrais**: Flujo de nuevas unidades de autoconsumo conectadas

In [18]:
centrais_df = pd.read_csv(r"../data_raw/26-centrais.csv", sep=';', encoding='utf-8')
centrais_df.head(5)

,ano,mes,con_name,cod_concelho,potencia_ligacao,processos_concluidos_nr
0,2024,11,Albufeira,801,10.96,3
1,2024,11,Elvas,1207,10.00,1
2,2024,11,Seixal,1510,15.57,6
3,2024,11,Loulé,808,110.00,1
4,2024,11,Setúbal,1512,12.00,1


In [19]:
centrais_df.shape

(77930, 6)

In [20]:
centrais_df.dtypes

ano                          int64
mes                          int64
con_name                    object
cod_concelho                 int64
potencia_ligacao           float64
processos_concluidos_nr      int64
dtype: object

In [ ]:
centrais_df.describe()

,ano,mes,cod_concelho,potencia_ligacao,processos_concluidos_nr
count,77930.000000,77930.000000,77930.000000,77134.000000,77930.000000
mean,2023.217426,6.663442,986.435121,18.998986,3.203323
std,1.302918,3.464444,524.184866,68.928357,5.949538
min,2021.000000,1.000000,101.000000,0.000000,1.000000
25%,2022.000000,4.000000,601.000000,2.000000,1.000000
50%,2023.000000,7.000000,1101.000000,4.960000,2.000000
75%,2024.000000,10.000000,1410.000000,14.300000,4.000000
max,2026.000000,12.000000,1824.000000,5406.000000,483.000000


In [22]:
centrais_df.isna().sum()

ano                          0
mes                          0
con_name                     0
cod_concelho                 0
potencia_ligacao           796
processos_concluidos_nr      0
dtype: int64

In [23]:
centrais_df.duplicated().sum()

np.int64(0)

In [24]:
centrais_df[['ano','mes','cod_concelho']].duplicated().sum()

np.int64(63288)

In [25]:
mask = centrais_df.duplicated(
    subset=['ano', 'mes', 'cod_concelho'],
    keep=False
)

centrais_df[mask].sort_values(
    ['ano', 'mes', 'cod_concelho']
).head(10)

,ano,mes,con_name,cod_concelho,potencia_ligacao,processos_concluidos_nr
13440,2021,7,Águeda,101,36.00,1
13441,2021,7,Águeda,101,2.70,4
13442,2021,7,Águeda,101,1.25,2
45050,2021,7,Águeda,101,25.00,1
45051,2021,7,Águeda,101,2.40,3
45267,2021,7,Águeda,101,0.85,2
68088,2021,7,Águeda,101,7.75,3
68155,2021,7,Águeda,101,0.00,1
68156,2021,7,Águeda,101,200.00,3
68366,2021,7,Águeda,101,0.75,1


In [65]:
centrais_df['data'] = pd.to_datetime(centrais_df[['ano','mes']].rename(
    columns={'ano': 'year', 'mes':'month'}).assign(day=1))

centrais_df.insert(0, 'data', centrais_df.pop('data'))

centrais_df['data'].min(), centrais_df['data'].max()

(Timestamp('2021-07-01 00:00:00'), Timestamp('2026-01-01 00:00:00'))

## Observaciones sobre el dataset '26-centrais'

Se confirman unidades correctas conforme metadatos del catálogo.

El dataset presenta 77930 filas y 6 columnas, siendo un punto a valorar a la hora de combinar datasets. La granularidad en este caso tampoco se marca a nivel de concelho, se ha comprobado que para el mismo año, mes y concelho hay muchas combinaciones repetidas, siendo 'potencia_ligacao' la dimensión adicional que completa la granularidad.

cod_concelho se utilizará como identificador territorial principal de cara a futuras agregaciones y merge entre datasets.

Las unidades del dataset son correctas, si bien falta crear una columna de tipo 'datetime' que combine los datos de año y mes para crear una fecha siguiendo la estructura de los anteriores datasets.

En cuanto a la distribución de las variables, parece que vuelve a existir un gran sesgo hacia valores elevados en lo relativo a la potencia de conexión y número de procesos (instalaciones) realizadas. Habrá que estudiar si mantiene coherencia o no.

Existen 796 nulos en cuanto a potencia de conexión, conviene investigarlos aún siendo un 1% de los registros totales.

En cuanto a la granularidad, queda marcada por año, mes y concelho, dando dos métricas, la potencia instalada y las unidades instaladas, es decir que en un futuro, cuando haya que combinar datasets hay que tener en cuenta si tomar ambas métricas o solo una, sabiendo que salen del sumatorio por grupo creado.

No hay duplicados generales contando todas las variables, en cuanto al rango de fechas, empieza en julio de 2021 y concluye en enero de 2026, reduciendo el rango dado que se acorta un poco más la última fecha disponible para todos los datasets por el momento.

## **comunidades-de-energia**: Número de comunidades de energias activas por tipo

In [27]:
comunidades_df = pd.read_csv(r"../data_raw/comunidades-de-energia.csv", sep=';', encoding='utf-8')
comunidades_df.head(5)

,data,ano,mes,dis_name,con_name,fre_name,tipo_acc_cer,contagem_tipo_acc_cer,codigo_distrito,codigo_concelho,codigo_freguesia
0,2025-09,2025,9,Braga,Braga,Adaúfe,ACC,3,3,303,030301
1,2025-09,2025,9,Porto,Santo Tirso,Água Longa,ACC,1,13,1314,131402
2,2025-09,2025,9,Lisboa,Amadora,Águas Livres,ACC,1,11,1115,111513
3,2025-09,2025,9,Santarém,Santarém,Alcanede,ACC,1,14,1416,141604
4,2025-09,2025,9,Lisboa,Lisboa,Alcântara,ACC,1,11,1106,110602


In [28]:
comunidades_df.shape

(11007, 11)

In [29]:
comunidades_df.dtypes

data                     object
ano                       int64
mes                       int64
dis_name                 object
con_name                 object
fre_name                 object
tipo_acc_cer             object
contagem_tipo_acc_cer     int64
codigo_distrito           int64
codigo_concelho           int64
codigo_freguesia         object
dtype: object

In [30]:
comunidades_df.isna().sum()

data                       0
ano                        0
mes                        0
dis_name                   0
con_name                   0
fre_name                 184
tipo_acc_cer               0
contagem_tipo_acc_cer      0
codigo_distrito            0
codigo_concelho            0
codigo_freguesia           0
dtype: int64

In [31]:
comunidades_df.duplicated().sum()

np.int64(0)

In [33]:
comunidades_df[['ano','mes','codigo_concelho','codigo_freguesia','tipo_acc_cer']].duplicated().sum()

np.int64(0)

In [34]:
comunidades_df['data'] = pd.to_datetime(comunidades_df['data'])

comunidades_df['data'].min(), comunidades_df['data'].max()

(Timestamp('2022-06-01 00:00:00'), Timestamp('2026-07-01 00:00:00'))

## Observaciones sobre el dataset 'comunidades-de-energia'

Se confirman correctas las unidades en las que se representan las variables.

El dataset consiste en 11.007 filas y 11 columnas. En cuanto a los tipos de datos, habrá que transformar data a datetime y tipo_acc_cer podría convertirse en variable categórica. Los códigos territoriales se mantendrán como identificadores y posteriormente podrán transformarse a string.

Los únicos valores nulos aparecen en fre_name (184 registros). Dado que codigo_freguesia está completo, estos nulos no impiden identificar territorialmente las observaciones.

En cuanto a la granularidad, una vez descartados los duplicados exactos, se comprueba que la combinación año + mes + codigo_concelho + codigo_freguesia + tipo_acc_cer no presenta duplicados, por lo que parece definir la granularidad original del dataset.

En futuros pasos probablemente será necesario agregar los datos a nivel de mes + concelho y, si interesa conservar el desglose por tipo de comunidad energética, realizar algún tipo de pivotado para transformar tipo_acc_cer en indicadores separados.

Finalmente, el dataset abarca desde junio de 2022 hasta julio de 2026. Presenta, por tanto, un periodo temporal ampliamente coincidente con los datasets anteriores, aunque antes de combinarlos habrá que homogeneizar su granularidad y las claves territoriales.

## **codigo-de-atividade-economica-por-distrito-concelho-e-freguesia**: Actividad economica por freguesia

In [35]:
atividade_economica_df = pd.read_csv(r"../data_raw/codigo-de-atividade-economica-por-distrito-concelho-e-freguesia.csv", sep=';', encoding='utf-8')
atividade_economica_df.head(5)

,ano,nuts_iii,codigo_cae,cae,no_de_empresas,dis_name,con_name,freguesia,codigo_nuts_iii,codigo_distrito,codigo_concelho,codigo_freguesia
0,2023,Alto Minho,F,Construção,6,Viana do Castelo,Arcos de Valdevez,Aboim das Choças,111,16,1601,160101
1,2023,Alto Minho,G,Comércio por grosso e a retalho; reparação de ...,7,Viana do Castelo,Arcos de Valdevez,Aboim das Choças,111,16,1601,160101
2,2023,Alto Minho,I,"Alojamento, restauração e similares",3,Viana do Castelo,Arcos de Valdevez,Aboim das Choças,111,16,1601,160101
3,2023,Alto Minho,P,Educação,3,Viana do Castelo,Arcos de Valdevez,Aboim das Choças,111,16,1601,160101
4,2023,Alto Minho,Q,Actividades de saúde humana e apoio social,5,Viana do Castelo,Arcos de Valdevez,Aboim das Choças,111,16,1601,160101


In [36]:
atividade_economica_df.shape

(39686, 12)

In [37]:
atividade_economica_df.dtypes

ano                  int64
nuts_iii            object
codigo_cae          object
cae                 object
no_de_empresas       int64
dis_name            object
con_name            object
freguesia           object
codigo_nuts_iii     object
codigo_distrito      int64
codigo_concelho      int64
codigo_freguesia    object
dtype: object

In [38]:
atividade_economica_df.isna().sum()

ano                    0
nuts_iii               0
codigo_cae             0
cae                    0
no_de_empresas         0
dis_name            2705
con_name               0
freguesia              0
codigo_nuts_iii        0
codigo_distrito        0
codigo_concelho        0
codigo_freguesia       0
dtype: int64

In [39]:
atividade_economica_df.duplicated().sum()

np.int64(0)

In [42]:
atividade_economica_df[['ano', 'codigo_freguesia', 'codigo_cae']].duplicated().sum()

np.int64(0)

In [ ]:
atividade_economica_df['ano'] = pd.to_datetime(atividade_economica_df['ano'], format='%Y')
atividade_economica_df['ano'].max()

Timestamp('1970-01-01 00:00:00.000002023')

In [44]:
atividade_economica_df['ano'].min(), atividade_economica_df['ano'].max()

(np.int64(2023), np.int64(2023))

## Observaciones sobre el dataset 'codigo-de-atividade-economica-por-distrito-concelho-e-freguesia'

Se confirman correctas las unidades en las que se representan las variables.

El dataset consiste en 39.686 filas y 12 columnas.

Es necesario cambiar codigo_cae a tipo categórico y también podría hacerse lo mismo con cae.

Los únicos valores nulos se presentan en dis_name. No parece necesario gestionarlos, ya que los códigos territoriales están completos y serán los utilizados para tareas de agrupación y futuras combinaciones entre datasets.

En cuanto a la granularidad, una vez descartados los duplicados globales, se comprueba que el dataset presenta una granularidad basada en año + codigo_freguesia + codigo_cae, siendo no_de_empresas la métrica.

De forma similar a los datasets anteriores, será necesario agregar posteriormente a nivel de concelho y posiblemente pivotar codigo_cae para mantener una estructura final donde cada combinación territorial y temporal ocupe una única fila.

Finalmente, se ha comprobado que el dataset de actividad económica solo cubre 2023. Los datos podrán incorporarse al dataset general mediante las claves geográficas, pero las hipótesis que utilicen estas variables se limitarán al periodo correspondiente a 2023, evitando extrapolar la estructura económica observada a otros años. Esta limitación temporal deberá tenerse en cuenta al interpretar los resultados.

## **clientes-por-escalao-de-potencia**: Número de contratos por segmento de potencia contratada

In [47]:
clientes_escalao_df = pd.read_csv(r"../data_raw/clientes-por-escalao-de-potencia.csv", sep=';', encoding='utf-8')
clientes_escalao_df.head(5)

,data,ano,mes,dis_name,con_name,fre_name,segmento_de_potencia_contratada,numero_de_contratos,dis_code,con_code,fre_code
0,2025-09,2025,9,Évora,Évora,Nossa Senhora da Tourega e Nossa Senhora de Gu...,]20 30],2.0,7,705,070525
1,2025-09,2025,9,Évora,Évora,Nossa Senhora da Tourega e Nossa Senhora de Gu...,]40 50],15.0,7,705,070525
2,2025-09,2025,9,Évora,Évora,Nossa Senhora da Tourega e Nossa Senhora de Gu...,]50 100],6.0,7,705,070525
3,2025-09,2025,9,Évora,Évora,Nossa Senhora da Tourega e Nossa Senhora de Gu...,]250 300],2.0,7,705,070525
4,2025-09,2025,9,Évora,Évora,Nossa Senhora da Tourega e Nossa Senhora de Gu...,]400 450],1.0,7,705,070525


In [48]:
clientes_escalao_df.shape

(2185853, 11)

In [49]:
clientes_escalao_df.dtypes

data                                object
ano                                  int64
mes                                  int64
dis_name                            object
con_name                            object
fre_name                            object
segmento_de_potencia_contratada     object
numero_de_contratos                float64
dis_code                             int64
con_code                             int64
fre_code                            object
dtype: object

In [ ]:
(clientes_escalao_df['numero_de_contratos'] % 1 == 0).all() # Calcula el resto de dividir cada valor entre 1, si todo son enteros debe dar 0

np.True_

In [53]:
clientes_escalao_df.isna().sum()

data                               0
ano                                0
mes                                0
dis_name                           0
con_name                           0
fre_name                           0
segmento_de_potencia_contratada    0
numero_de_contratos                0
dis_code                           0
con_code                           0
fre_code                           0
dtype: int64

In [50]:
clientes_escalao_df.duplicated().sum()

np.int64(0)

In [51]:
clientes_escalao_df[['ano','mes','fre_code','segmento_de_potencia_contratada']].duplicated().sum()

np.int64(0)

In [52]:
clientes_escalao_df['data'] = pd.to_datetime(clientes_escalao_df['data'])

clientes_escalao_df['data'].min(), clientes_escalao_df['data'].max()

(Timestamp('2023-01-01 00:00:00'), Timestamp('2026-07-01 00:00:00'))

## Observaciones sobre el dataset 'clientes-por-escalao-de-potencia'

Se confirman las unidades presentadas para cada variable en el catálogo.

El dataset presenta 2185853 filas y 11 columnas; es lógico dado que hay una gran variedad de segmentos de potencia contratada y se calculan los contratos por freguesia.

El tipo de dato debe ser cambiado para 'data' a datetime, como en anteriores casos. 'segmento_de_potencia_contratada' debería ser categórica ordinal y 'numero_de_contratos' tiene más sentido que sea int64 una vez comprobado que efectivamente solo existen enteros para esa dimensión.
El resto de variables se pueden mantener igual si bien todos los códigos al ser identificadores encajan mejor como 'string'.

No hay duplicados generales y parece que la granularidad se reduce a año, mes, código de freguesia (menor unidad territorial) y segmento de potencia contratada. En este caso la métrica son el número de contratos.

El rango de fechas abarca desde enero de 2023 a julio de 2026, presentando una amplia coincidencia temporal con el resto de datasets.

## **25-plr-producao-renovavel**: Nuevas conexiones a la red de centrales de energía renovable

In [55]:
producao_renovavel_df = pd.read_csv(r"../data_raw/25-plr-producao-renovavel.csv", sep=';', encoding='utf-8')
producao_renovavel_df.head(5)

,ano,mes,con_name,cod_concelho,potencia_ligacao,pedidos_ligacao_rede_executados_nr
0,2023,12,Santarém,1416,3000,3
1,2024,8,Santarém,1416,3000,3
2,2025,3,Santarém,1416,1000,1
3,2021,6,Salvaterra de Magos,1415,25000,1
4,2025,11,Salvaterra de Magos,1415,1000,1


In [56]:
producao_renovavel_df.shape

(253, 6)

In [57]:
producao_renovavel_df.dtypes

ano                                    int64
mes                                    int64
con_name                              object
cod_concelho                           int64
potencia_ligacao                       int64
pedidos_ligacao_rede_executados_nr     int64
dtype: object

In [58]:
producao_renovavel_df.isna().sum()

ano                                   0
mes                                   0
con_name                              0
cod_concelho                          0
potencia_ligacao                      0
pedidos_ligacao_rede_executados_nr    0
dtype: int64

In [60]:
producao_renovavel_df.duplicated().sum()

np.int64(0)

In [62]:
producao_renovavel_df[['ano','mes','cod_concelho','potencia_ligacao']].duplicated().sum()

np.int64(0)

In [66]:
producao_renovavel_df['data'] = pd.to_datetime(producao_renovavel_df[['ano', 'mes']].rename(columns={'ano': 'year', 'mes': 'month'}).assign(day=1))
producao_renovavel_df.insert(0,'data',producao_renovavel_df.pop('data'))

producao_renovavel_df['data'].min(), producao_renovavel_df['data'].max()

(Timestamp('2021-02-01 00:00:00'), Timestamp('2026-01-01 00:00:00'))

## Observaciones sobre el dataset '25-plr-producao-renovavel'

La unidades encajan con lo indicado en el catálogo.

Se trata de un dataset bastante reducido de apenas 253 filas y 6 columnas.
En cuanto a tipos, cabría transformar a texto el código de concelho y crear una variable 'data' que combine año y mes.

No hay valores nulos ni duplicados globales.
La granularidad es por año y mes así como concelho y potencia instalada. La métrica en este caso son las instalaciones ejecutadas.

El rango temporal abarca desde febrero de 2021 hasta enero de 2026, por lo que presenta una amplia coincidencia temporal con los datasets anteriores.